### PR TITLE
[Test] unit test checking cold staking script key extraction.

### PR DESCRIPTION
**Pull request by furszy**
Essentially proving #1830 bug and #1888 subsequent fix.

### CODE REVIEW

File: src/test/coldstaking_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+BOOST_FIXTURE_TEST_SUITE(coldstaking_tests, BasicTestingSetup)
+
+// Goal: check cold staking script keys extraction
+BOOST_AUTO_TEST_CASE(extract_cold_staking_destination_keys)
+{
+    CKey ownerKey;
+    ownerKey.MakeNewKey(true);
+    CKey stakerKey;
+    stakerKey.MakeNewKey(true);
+    CScript script = GetScriptForStakeDelegation(ownerKey.GetPubKey().GetID(), stakerKey.GetPubKey().GetID());
```
Comment (random-zebra) [inline src/test/coldstaking_tests.cpp:20]: This is switched, the first argument is the staker's key, the second is the owner's.
```cpp
CScript script = GetScriptForStakeDelegation(stakerKey.GetPubKey().GetID(), ownerKey.GetPubKey().GetID());
```

File: src/test/coldstaking_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Check owner
+    CTxDestination ownerDest;
+    BOOST_CHECK(ExtractDestination(script, ownerDest, false));
+    BOOST_CHECK(boost::get<CKeyID>(&ownerDest));
```
Comment (random-zebra) [inline src/test/coldstaking_tests.cpp:25]: This only checks that the destination is valid (and we should use `IsValidDestination()` for that).
Would be good to compare it to the expected owner's key:
```cpp
    BOOST_CHECK(IsValidDestination(ownerDest));
    BOOST_CHECK(*boost::get<CKeyID>(&ownerDest) == ownerKey.GetPubKey().GetID());
```

File: src/test/coldstaking_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Check staker
+    CTxDestination stakerDest;
+    BOOST_CHECK(ExtractDestination(script, stakerDest, true));
+    BOOST_CHECK(boost::get<CKeyID>(&stakerDest));
```
Comment (random-zebra) [inline src/test/coldstaking_tests.cpp:30]: Here too. Let's compare against the expected staker's key:
```cpp
    BOOST_CHECK(IsValidDestination(stakerDest));
    BOOST_CHECK(*boost::get<CKeyID>(&stakerDest) == stakerKey.GetPubKey().GetID());
```

File: src/test/coldstaking_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    int nRequiredRet = -1;
+    std::vector<CTxDestination> destVector;
+    BOOST_CHECK(ExtractDestinations(script, type, destVector, nRequiredRet));
+    BOOST_CHECK(type == TX_COLDSTAKE);
+    BOOST_CHECK(nRequiredRet == 2);
+    BOOST_CHECK(destVector.size() == 2);
+
+    for (const auto& dest : destVector) {
+        BOOST_CHECK(boost::get<CKeyID>(&dest));
+    }
```
Comment (random-zebra) [inline src/test/coldstaking_tests.cpp:43]: Here we can just check against the results of `ExtractDestination`:
```cpp
    BOOST_CHECK(destVector[0] == stakerDest);
    BOOST_CHECK(destVector[1] == ownerDest);
```

**Review by random-zebra — COMMENTED** [review]
Nice. 
It would be good to also check the returned public key hashes (instead of only verifying that the destinations are valid).

I would also suggest to rename this file to  `script_P2CS_tests.cpp` (similar to `script_P2SH_tests`) as it is meant to test specifically the script encode/decode stuff of the cold staking protocol.

**Comment by furszy** [comment]
Updated per feedback. 
With a minor thing that haven't added. Not using `IsValidDestination` because what we need to ensure is that the destination is a CKeyID and nothing else.

**Comment by random-zebra** [comment]
Right now the test crashes for a memory access violation, instead of gracefully throwing a BOOST_CHECK error.
Imo, it would be better to refactor the check as something like:
```cpp
void CheckValidKeyId(const CTxDestination& dest, const CKeyID& expectedKey)
{
    if (dest.which() == 1) {
        const CKeyID* keyid = boost::get<CKeyID>(&dest);
        BOOST_CHECK(keyid);
        BOOST_CHECK(*keyid == expectedKey);
    } else {
        BOOST_ERROR("Destination is not a CKeyID");
    }
}
```

**Comment by furszy** [comment]
upps yeah, too many things going on hehe. Fixed.

**Review by Fuzzbawls — APPROVED** [review]
ACK b189cc37c6d295a4c855d19806068e810736f736

**Review by random-zebra — APPROVED** [review]
ACK b189cc37c6d295a4c855d19806068e810736f736 (failing on master, passing on #1888). Merging...

**Pull request merged**